README.md
# docker-dns

##### Run a dns proxy in docker with wildcard host support

This is intended to be used within (but not limited to) a local docker development environment, typically in combination with some type of reverse proxy that supports virtualhosts like nginx. In this configuration, `docker-dns` allows you to simplify local container access by automatically setting up easier doamins like `hello.docker`

#### Quickstart:

`docker run -p 53:53/udp -d --restart always finboxio/docker-dns`

Once this container is running, configure your system to route DNS queries through localhost and you're all set.

#### Configuration:

`docker-dns` can be configured through the use of a few environment variables:

-  `-e PORT=${PORT:-53}` sets the port the DNS server will listen on
-  `-e EXTERNAL_DNS=${IP:-8.8.8.8}` sets the DNS server to which queries that cannot be resolved locally are proxied
-  `-e HOSTS_FILE=${FILE:-/etc/docker-dns/hosts} -v $FILE:$FILE` sets the host file to read for resolving queries locally. This should look like a typical /etc/hosts file, but additionally supports wildcard entries. The default hosts file included in this image is shown below, but you are free to mount your own:

	```
	127.0.0.1 		*.docker
	192.168.99.100 	*.rancher
	```

> Note:
>
> The second entry is really only effective if you have a rancher host running in a virtual machine with the IP 192.168.99.100. I do some development against a local rancher cluster spun up via [mac-ranch](https://github.com/finboxio/mac-ranch), so this is actually a reasonably useful default for me.
>

#### Using with a reverse proxy

[jwilder/nginx-proxy](https://github.com/jwilder/nginx-proxy) makes this really easy:

`docker run -d -p 80:80 -v /var/run/docker.sock:/tmp/docker.sock:ro --restart always jwilder/nginx-proxy`

> Note:
>
> Your system may have a default web server already running on port 80. If that's the case you'll either need to disable it or choose a different port. If you're on mac and aren't using the default apache server you can disable it with
>
> `sudo launchctl unload -w /System/Library/LaunchDaemons/org.apache.httpd.plist`
>

Now, you can start up any container with `VIRTUAL_HOST` set to a `docker` domain and access it from the browser without having to expose/keep track of all the different ports you're using!

`docker run -d -p 80 -e VIRTUAL_HOST=hello.docker -e VIRTUAL_PORT=80 tutum/hello-world && open http://hello.docker`

#### \#winning

#### Acknowlegements

DNS proxy code shamelessly copied from [hubdotcom/marlon-tools](https://github.com/hubdotcom/marlon-tools). Thanks for the great work!


History.md

0.1.0 / 2016-11-30
==================

  * switch to more robust proxy written in node

0.0.5 / 2016-09-11
==================

  * remove .dev default tld

0.0.4 / 2016-09-11
==================

  * shorten ttl to 1, disable caching by default, add better logging

0.0.3 / 2016-09-10
==================

  * update defaults to .dev

0.0.2 / 2016-09-10
==================

  * add reference to original dns code

0.0.1 / 2016-09-10
==================

  * Add readme

0.0.0 / 2016-09-10
==================

  * Initial commit


index.js
var PORT = process.env.PORT || 53
var EXTERNAL_DNS = process.env.EXTERNAL_DNS || '8.8.8.8,8.8.4.4'

var opts = require('rc')('dnsproxy', {
  host: '0.0.0.0',
  logging: 'dnsproxy:query',
  domains: {
    '.docker': '127.0.0.1',
    '.rancher': '192.168.99.100'
  },
  fallback_timeout: 350
})

if (!opts.port) opts.port = PORT
if (!opts.host) opts.host = '0.0.0.0'
if (!opts.nameservers) opts.nameservers = EXTERNAL_DNS.split(',')
if (!opts.servers) opts.servers = {}
if (!opts.domains) opts.domains = {}
if (!opts.hosts) opts.hosts = {}

process.env.DEBUG_FD = process.env.DEBUG_FD || 1
process.env.DEBUG = process.env.DEBUG || opts.logging
var d = process.env.DEBUG.split(',')
d.push('dnsproxy:error')
process.env.DEBUG = d.join(',')

var dgram = require('dgram')
var packet = require('native-dns-packet')
var util = require('./util.js')

var logdebug = require('debug')('dnsproxy:debug')
var logquery = require('debug')('dnsproxy:query')
var logerror = require('debug')('dnsproxy:error')

logdebug('options: %j', opts)

var server = dgram.createSocket('udp4')

server.on('listening', () => {
  var address = server.address();
  console.log(`dns server listening ${address.address}:${address.port}`);
});

server.on('error', function (err) {
  logerror('Server Error: %s', err)
})

server.on('message', function (message, rinfo) {
  var nameserver = opts.nameservers[0]
  var returner = false

  var query = packet.parse(message)
  var domain = query.question[0].name
  var type = query.question[0].type

  logdebug('query: %j', query)

  Object.keys(opts.hosts).forEach(function (h) {
    if (domain === h) {
      var answer = opts.hosts[h]
      if (typeof opts.hosts[opts.hosts[h]] !== 'undefined') {
        answer = opts.hosts[opts.hosts[h]]
      }

      logquery('type: host, domain: %s, answer: %s', domain, opts.hosts[h])

      var res = util.createAnswer(query, answer)
      server.send(res, 0, res.length, rinfo.port, rinfo.address)

      returner = true
    }
  })

  if (returner) {
    return
  }

  Object.keys(opts.domains).forEach(function (s) {
    var sLen = s.length
    var dLen = domain.length

    if (domain.indexOf(s) >= 0 && domain.indexOf(s) === (dLen - sLen)) {
      var answer = opts.domains[s]
      if (typeof opts.domains[opts.domains[s]] !== 'undefined') {
        answer = opts.domains[opts.domains[s]]
      }

      logquery('type: server, domain: %s, answer: %s', domain, opts.domains[s])

      var res = util.createAnswer(query, answer)
      server.send(res, 0, res.length, rinfo.port, rinfo.address)

      returner = true
    }
  })

  if (returner) {
    return
  }

  Object.keys(opts.servers).forEach(function (s) {
    if (domain.indexOf(s) !== -1) {
      nameserver = opts.servers[s]
    }
  })

  var fallback
  (function queryns (message, nameserver) {
    var sock = dgram.createSocket('udp4')
    sock.send(message, 0, message.length, 53, nameserver, function () {
      fallback = setTimeout(function () {
        queryns(message, opts.nameservers[0])
      }, opts.fallback_timeout)
    })
    sock.on('error', function (err) {
      logerror('Socket Error: %s', err)
      process.exit(5)
    })
    sock.on('message', function (response) {
      clearTimeout(fallback)
      logquery('type: primary, nameserver: %s, query: %s, type: %s, answer: %s', nameserver, domain, util.records[type] || 'unknown', util.listAnswer(response))
      server.send(response, 0, response.length, rinfo.port, rinfo.address)
      sock.close()
    })
  }(message, nameserver))
})

server.bind(opts.port, opts.host)



dnsparser.py
# coding: utf-8
from cStringIO import StringIO
import struct

DNS_FLAG_QR = 0x8000
DNS_FLAG_RD = 0x0100
DNS_FLAG_RA = 0x0080

DNS_TYPE_A = 1
DNS_TYPE_AAAA = 28
DNS_CLASS_IN = 1

'''
一个简单的DNS解析器。

解析：
msg = DNSMessage.parse(message data)

序列化：
data = msg.serialize()


author: marlonyao<yaolei135@gmail.com>
'''
class DNSMessageHeader(object):
    def __init__(self, id, flag, qd_count, an_count, ns_count, ar_count):
        self.id = id
        self.flag = flag
        self.qd_count = qd_count
        self.an_count = an_count
        self.ns_count = ns_count
        self.ar_count = ar_count

    @staticmethod
    def parse(message):
        id, flag = struct.unpack('!HH', message.read(4))
        qd_count, an_count, ns_count, ar_count = struct.unpack('!HHHH', message.read(8))
        return DNSMessageHeader(id, flag, qd_count, an_count, ns_count, ar_count)

    def serialize(self, message, memoize=None):
        message.write(struct.pack('!HHHHHH', self.id, self.flag,
                       self.qd_count, self.an_count, self.ns_count, self.ar_count))

    def __str__(self):
        return 'id: %s, flag: %s, qd_count: %s, an_count: %s, ns_count: %s, ar_count: %s' % (
            self.id, self.flag, self.qd_count, self.an_count, self.ns_count, self.ar_count
        )

class DNSMessageQuestion(object):
    def __init__(self, qname, qtype, qclass):
        self.qname = qname
        self.qtype = qtype
        self.qclass = qclass

    @staticmethod
    def parse(message):
        qname = parse_domain_name(message)
        qtype, qclass = struct.unpack('!HH', message.read(4))
        return DNSMessageQuestion(qname, qtype, qclass)

    def serialize(self, message, memoize):
        unparse_domain_name(self.qname, message, memoize)
        message.write(struct.pack('!HH', self.qtype, self.qclass))

    def __str__(self):
        return 'qname: %s, qtype: %s, qclass: %s' % (self.qname, self.qtype, self.qclass)

    def __repr__(self):
        return str(self)

class DNSMessageRecord(object):
    def __init__(self, name, type_, class_, ttl, rdata):
        self.name = name
        self.type_ = type_
        self.class_ = class_
        self.ttl = ttl
        self.rdata = rdata

    @staticmethod
    def parse(message):
        name = parse_domain_name(message)
        type_, class_, ttl, rd_len = struct.unpack('!HHIH', message.read(10))
        return DNSMessageRecord(name, type_, class_, ttl, message.read(rd_len))

    def serialize(self, message, memoize):
        unparse_domain_name(self.name, message, memoize)
        message.write(struct.pack('!HHIH%ss'%len(self.rdata),
                    self.type_, self.class_, self.ttl,
                    len(self.rdata), self.rdata,
                ))

    def __str__(self):
        return "name: %s, type: %s, class: %s, ttl: %s, rdata: %s" % (
            self.name, self.type_, self.class_, self.ttl, self.rdata
        )

    def __repr__(self):
        return str(self)

class DNSMessage(object):
    def __init__(self, header, questions=None, answers=None, authorities=None, additionals=None):
        self.header = header
        self.questions = questions or []
        self.answers = answers or []
        self.authorities = authorities or []
        self.additionals = additionals or []

    @staticmethod
    def parse(data):
        message = StringIO(data)
        header = DNSMessageHeader.parse(message)
        questions = []
        for i in range(0, header.qd_count):
            quest = DNSMessageQuestion.parse(message)
            questions.append(quest)
        answers = []
        for i in range(0, header.an_count):
            answer = DNSMessageRecord.parse(message)
            answers.append(answer)
        authorities = []
        for i in range(0, header.ns_count):
            authority = DNSMessageRecord.parse(message)
            authorities.append(authority)
        additionals = []
        for i in range(0, header.ar_count):
            additional = DNSMessageRecord.parse(message)
            additionals.append(additional)
        return DNSMessage(header, questions, answers, authorities, additionals)

    def serialize(self):
        'serialize to network bytes, not considering name compression'
        message = StringIO()
        memoize = {}
        self.header.serialize(message, memoize)
        for s in self.questions:
            s.serialize(message, memoize)
        for s in self.answers:
            s.serialize(message, memoize)
        for s in self.authorities:
            s.serialize(message, memoize)
        for s in self.additionals:
            s.serialize(message, memoize)
        return message.getvalue()

    def __str__(self):
        return 'header: %s\n' % self.header +\
               'questions: %s\n' % self.questions +\
               'answers: %s\n' % self.answers +\
               'authorities: %s\n' % self.authorities +\
               'additionals: %s\n' % self.additionals

    def __repr__(self):
        return str(self)

def _parse_domain_labels(message):
    labels = []
    len = ord(message.read(1))
    while len > 0:
        if len >= 64:   # domain name compression
            len = len & 0x3f
            offset = (len << 8) + ord(message.read(1))
            mesg = StringIO(message.getvalue())
            mesg.seek(offset)
            labels.extend(_parse_domain_labels(mesg))
            return labels
        else:
            labels.append(message.read(len))
            len = ord(message.read(1))
    return labels
def parse_domain_name(message):
    return '.'.join(_parse_domain_labels(message))

def unparse_domain_name(name, message, memoize):
    labels = name.split('.')
    for i, label in enumerate(labels):
        qname = '.'.join(labels[i:])
        if qname in memoize:
            offset = (memoize[qname] & 0x3fff) + 0xc000
            message.write(struct.pack('!H', offset))
            break
        else:
            memoize[qname] = message.tell()
            #print 'add to memoize: %s, %s' % (qname, message.tell())
            message.write(struct.pack('!B%ss' % len(label), len(label), label))
    else:
        # write last ending zero
        message.write('\x00')


dnsproxy.py
#!/usr/bin/env python
# coding: utf-8
from SocketServer import BaseRequestHandler, ThreadingUDPServer
from cStringIO import StringIO
import os
import socket
import struct
import time

'''
A simple DNS proxy server, support wilcard hosts, IPv6, cache. Usage:

Edit /etc/hosts, add:
    127.0.0.1 *.local
    2404:6800:8005::62 *.blogspot.com
startup dnsproxy(here use Google DNS server as delegating server):
$ sudo python dnsproxy.py -s 8.8.8.8

Then set system dns server as 127.0.0.1, you can verify it by dig:
$ dig test.local

The result should contains 127.0.0.1.

author: marlonyao<yaolei135@gmail.com>
'''
def main():
    import optparse, sys
    parser = optparse.OptionParser()
    parser.add_option('-f', '--hosts-file', dest='hosts_file', metavar='<file>', default='/etc/hosts', help='specify hosts file, default /etc/hosts')
    parser.add_option('-H', '--host', dest='host', default='127.0.0.1', help='specify the address to listen on')
    parser.add_option('-p', '--port', dest='port', default=53, type='int', help='specify the port to listen on')
    parser.add_option('-s', '--server', dest='dns_server', metavar='SERVER', help='specify the delegating dns server')
    parser.add_option('-C', '--no-cache', dest='disable_cache', default=True, action='store_true', help='disable dns cache')

    opts, args = parser.parse_args()
    if not opts.dns_server:
        parser.print_help()
        sys.exit(1)
    dnsserver = DNSProxyServer(opts.dns_server, disable_cache=opts.disable_cache, host=opts.host, port=opts.port, hosts_file=opts.hosts_file)
    dnsserver.serve_forever()

class Struct(object):
    def __init__(self, **kwargs):
        for name, value in kwargs.items():
            setattr(self, name, value)

def parse_dns_message(data):
    message = StringIO(data)
    message.seek(4)     # skip id, flag
    c_qd, c_an, c_ns, c_ar = struct.unpack('!4H', message.read(8))
    # parse question
    question = parse_dns_question(message)
    for i in range(1, c_qd):    # skip other question
        parse_dns_question(message)
    records = []
    for i in range(c_an+c_ns+c_ar):
        records.append(parse_dns_record(message))
    return Struct(question=question, records=records)

def parse_dns_question(message):
    qname = parse_domain_name(message)
    qtype, qclass = struct.unpack('!HH', message.read(4))
    end_offset = message.tell()
    return Struct(name=qname, type_=qtype, class_=qclass, end_offset=end_offset)

def parse_dns_record(message):
    parse_domain_name(message)      # skip name
    message.seek(4, os.SEEK_CUR)    # skip type, class
    ttl_offset = message.tell()
    ttl = struct.unpack('!I', message.read(4))[0]
    rd_len = struct.unpack('!H', message.read(2))[0]
    message.seek(rd_len, os.SEEK_CUR)     # skip rd_content
    return Struct(ttl_offset=ttl_offset, ttl=ttl)

def _parse_domain_labels(message):
    labels = []
    len = ord(message.read(1))
    while len > 0:
        if len >= 64:   # domain name compression
            len = len & 0x3f
            offset = (len << 8) + ord(message.read(1))
            mesg = StringIO(message.getvalue())
            mesg.seek(offset)
            labels.extend(_parse_domain_labels(mesg))
            return labels
        else:
            labels.append(message.read(len))
            len = ord(message.read(1))
    return labels

def parse_domain_name(message):
    return '.'.join(_parse_domain_labels(message))

def addr_p2n(addr):
    try:
        return socket.inet_pton(socket.AF_INET, addr)
    except:
        return socket.inet_pton(socket.AF_INET6, addr)

DNS_TYPE_A = 1
DNS_TYPE_AAAA = 28
DNS_CLASS_IN = 1

class DNSProxyHandler(BaseRequestHandler):
    def handle(self):
        reqdata, sock = self.request
        req = parse_dns_message(reqdata)
        q = req.question
        print 'requesting', q.name
        if q.type_ in (DNS_TYPE_A, DNS_TYPE_AAAA) and (q.class_ == DNS_CLASS_IN):
            for packed_ip, host in self.server.host_lines:
                if q.name.endswith(host):
                    print 'returning', chr(len(packed_ip)), packed_ip, host
                    # header, qd=1, an=1, ns=0, ar=0
                    rspdata = reqdata[:2] + '\x81\x80\x00\x01\x00\x01\x00\x00\x00\x00'
                    rspdata += reqdata[12:q.end_offset]
                    # answer
                    rspdata += '\xc0\x0c'   # pointer to domain name
                    # type, 1 for ip4, 28 for ip6
                    if len(packed_ip) == 4:
                        rspdata += '\x00\x01'   # 1 for ip4
                    else:
                        rspdata += '\x00\x1c'   # 28 for ip6
                    # class: 1, ttl: 1(0x00000001)
                    rspdata += '\x00\x01\x00\x00\x00\x01'
                    rspdata += '\x00' + chr(len(packed_ip)) # rd_len
                    rspdata += packed_ip
                    sock.sendto(rspdata, self.client_address)
                    return

        # lookup cache
        if not self.server.disable_cache:
            cache = self.server.cache
            cache_key = (q.name, q.type_, q.class_)
            cache_entry = cache.get(cache_key)
            if cache_entry:
                rspdata = update_ttl(reqdata, cache_entry)
                if rspdata:
                    sock.sendto(rspdata, self.client_address)
                    return
        rspdata = self._get_response(reqdata)
        if not self.server.disable_cache:
            cache[cache_key] = Struct(rspdata=rspdata, cache_time=int(time.time()))
        sock.sendto(rspdata, self.client_address)

    def _get_response(self, data):
        sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM) # socket for the remote DNS server
        sock.connect((self.server.dns_server, 53))
        sock.sendall(data)
        sock.settimeout(60)
        rspdata = sock.recv(65535)
        sock.close()
        return rspdata

def update_ttl(reqdata, cache_entry):
    rspdata, cache_time = cache_entry.rspdata, cache_entry.cache_time
    rspbytes = bytearray(rspdata)
    rspbytes[:2] = reqdata[:2]          # update id
    current_time = int(time.time())
    time_interval = current_time - cache_time
    rsp = parse_dns_message(rspdata)
    for record in rsp.records:
        if record.ttl <= time_interval:
            return None
        rspbytes[record.ttl_offset:record.ttl_offset+4] = struct.pack('!I', record.ttl-time_interval)
    return str(rspbytes)

def load_hosts(hosts_file):
    'load hosts config, only extract config line contains wildcard domain name'
    print "loading hosts from file " + hosts_file
    def wildcard_line(line):
        parts = line.strip().split()[:2]
        if len(parts) < 2: return False
        if not parts[1].startswith('*'): return False
        try:
            packed_ip = addr_p2n(parts[0])
            return packed_ip, parts[1][1:]
        except:
            return None
    with open(hosts_file) as hosts_in:
        hostlines = []
        for line in hosts_in:
            hostline = wildcard_line(line)
            if hostline:
                print "adding host " , hostline
                hostlines.append(hostline)
        return hostlines

class DNSProxyServer(ThreadingUDPServer):
    def __init__(self, dns_server, disable_cache=False, host='127.0.0.1', port=53, hosts_file='/etc/hosts'):
        self.dns_server = dns_server
        self.hosts_file = hosts_file
        self.host_lines = load_hosts(hosts_file)
        self.disable_cache = disable_cache
        self.cache = {}
        ThreadingUDPServer.__init__(self, (host, port), DNSProxyHandler)

if __name__ == '__main__':
    main()
